File: shared/lib/extract.ts
/**
 * Body content extractors — run once at ingest (and during index migration);
 * results are stored as small index fields so the inbox can filter without
 * touching bodies.
 */
const CODE_RE = /\d{6}/g;
const LINK_RE = /https?:\/\/[^\s<>"'）)\]]+/gi;

/**
 * A 6-digit number only counts as a verification code when a code keyword —
 * 验证码/校验码/动态码 (CN) or code/verification/otp/passcode... (EN) — appears
 * within ±32 characters of it in the tag-stripped text. Bare numbers (order
 * ids, prices, timestamps) don't count.
 */
const CODE_HINT_RE =
    /(验证码|校验码|动态码|动态密码|认证码|短信码|验证代码)|(\b(verification|verify|verified|code|otp|one[ -]?time|passcode|pass[ -]?code|security|pin|confirm)\b)/i;

function stripHtml(html: string): string {
    return html
        .replace(/<style[\s\S]*?<\/style>/gi, " ")
        .replace(/<script[\s\S]*?<\/script>/gi, " ")
        .replace(/<[^>]+>/g, " ")
        .replace(/&nbsp;/gi, " ");
}

function plainSource(text?: string, html?: string): string {
    return `${text || ""}\n${stripHtml(html || "")}`.replace(/\s+/g, " ");
}

export function extractCodes(text?: string, html?: string): string[] {
    const plain = plainSource(text, html);
    if (!plain.trim()) return [];
    const out: string[] = [];
    for (const m of plain.matchAll(CODE_RE)) {
        const at = m.index ?? 0;
        const window = plain.slice(Math.max(0, at - 32), Math.min(plain.length, at + m[0].length + 32));
        if (!CODE_HINT_RE.test(window)) continue;
        out.push(m[0]);
        if (out.length >= 3) break;
    }
    return Array.from(new Set(out));
}

/** Distinct http(s) URLs found in the body (raw html kept — href URLs count), capped. */
export function extractLinks(text?: string, html?: string): string[] {
    const source = `${text || ""}\n${html || ""}`;
    if (!source.trim()) return [];
    const matches = source.match(LINK_RE) || [];
    const normalized = matches.map((u) => u.replace(/[.,;:!?]+$/, ""));
    return Array.from(new Set(normalized)).slice(0, 10);
}

File: server/extract.go
package main

import (
	"regexp"
	"strings"
)

// Port of shared/lib/extract.ts — verification-code + link flags recorded at ingest.

var (
	codeRe     = regexp.MustCompile(`\d{6}`)
	linkRe     = regexp.MustCompile(`(?i)https?://[^\s<>"'）)\]]+`)
	codeHintRe = regexp.MustCompile(`(验证码|校验码|动态码|动态密码|认证码|短信码|验证代码)|(\b(verification|verify|verified|code|otp|one[ -]?time|passcode|pass[ -]?code|security|pin|confirm)\b)`)
)

func stripHtmlTags(html string) string {
	s := regexp.MustCompile(`(?is)<style[\s\S]*?</style>`).ReplaceAllString(html, " ")
	s = regexp.MustCompile(`(?is)<script[\s\S]*?</script>`).ReplaceAllString(s, " ")
	s = regexp.MustCompile(`<[^>]+>`).ReplaceAllString(s, " ")
	return strings.ReplaceAll(s, "&nbsp;", " ")
}

func plainSource(text, html string) string {
	src := text + "\n" + stripHtmlTags(html)
	return regexp.MustCompile(`\s+`).ReplaceAllString(src, " ")
}

func extractCodes(text, html string) []string {
	plain := plainSource(text, html)
	if strings.TrimSpace(plain) == "" {
		return []string{}
	}
	seen := map[string]bool{}
	out := []string{}
	for _, loc := range codeRe.FindAllStringIndex(plain, -1) {
		start := loc[0] - 32
		if start < 0 {
			start = 0
		}
		end := loc[1] + 32
		if end > len(plain) {
			end = len(plain)
		}
		if !codeHintRe.MatchString(plain[start:end]) {
			continue
		}
		code := plain[loc[0]:loc[1]]
		if !seen[code] {
			seen[code] = true
			out = append(out, code)
		}
		if len(out) >= 3 {
			break
		}
	}
	if out == nil {
		out = []string{}
	}
	return out
}

func extractLinks(text, html string) []string {
	source := text + "\n" + html
	if strings.TrimSpace(source) == "" {
		return []string{}
	}
	seen := map[string]bool{}
	out := []string{}
	for _, m := range linkRe.FindAllString(source, -1) {
		m = strings.TrimRight(m, ".,;:!?")
		if !seen[m] {
			seen[m] = true
			out = append(out, m)
		}
	}
	if len(out) > 10 {
		out = out[:10]
	}
	if out == nil {
		out = []string{}
	}
	return out
}
